### PR TITLE
Use strings.TrimSuffix instead of strings.TrimRight to omit internal rune comparison function allocation

### PR DESCRIPTION
**Pull request by moredure**

### CODE REVIEW

**Review by temoto — APPROVED** [review]
LGTM. Could you post benchcmp for history?

**Comment by moredure** [comment]
Yeap, sure, @temoto, is it possible that more than one asterisk might happen in the end of line?

**Comment by moredure** [comment]
```go
func BenchmarkTrimSuffix(b *testing.B) {
	const line = "Allow: *"
	b.Run("HasSuffix+TrimRight", func(b *testing.B) {
		b.ReportAllocs()
		for i := 0; i < b.N; i++ {
			if strings.HasSuffix(line, "*") {
				_ = strings.TrimRight(line, "*")
			}
		}
	})
	b.Run("HasSuffix+TrimRightFunc", func(b *testing.B) {
		b.ReportAllocs()
		for i := 0; i < b.N; i++ {
			if strings.HasSuffix(line, "*") {
				_ = strings.TrimRightFunc(line, func (r rune) bool {
					return r == '*'
				})
			}
		}
	})
	b.Run("TrimRightFunc", func(b *testing.B) {
		b.ReportAllocs()
		for i := 0; i < b.N; i++ {
			_ = strings.TrimRightFunc(line, func (r rune) bool {
				return r == '*'
			})
		}
	})
	b.Run("TrimSuffix", func(b *testing.B) {
		b.ReportAllocs()
		for i := 0; i < b.N; i++ {
			_ = strings.TrimSuffix(line, "*")
		}
	})

	b.Run("TrimSuffixManually", func(b *testing.B) {
		b.ReportAllocs()
		for i := 0; i < b.N; i++ {
			if strings.HasSuffix(line, "*") {
				_ = line[:len(line)-1]
			}
		}
	})
}
```
```
cpu: Intel(R) Core(TM) i5-7600 CPU @ 3.50GHz
BenchmarkTrimSuffix
BenchmarkTrimSuffix/HasSuffix+TrimRight
BenchmarkTrimSuffix/HasSuffix+TrimRight-4         	24147571	        45.49 ns/op	      24 B/op	       1 allocs/op
BenchmarkTrimSuffix/HasSuffix+TrimRightFunc
BenchmarkTrimSuffix/HasSuffix+TrimRightFunc-4     	100000000	        11.85 ns/op	       0 B/op	       0 allocs/op
BenchmarkTrimSuffix/TrimRightFunc
BenchmarkTrimSuffix/TrimRightFunc-4               	100000000	        11.19 ns/op	       0 B/op	       0 allocs/op
BenchmarkTrimSuffix/TrimSuffix
BenchmarkTrimSuffix/TrimSuffix-4                  	1000000000	         0.2609 ns/op	       0 B/op	       0 allocs/op
BenchmarkTrimSuffix/TrimSuffixManually
BenchmarkTrimSuffix/TrimSuffixManually-4          	1000000000	         0.2520 ns/op	       0 B/op	       0 allocs/op
```

**Comment by temoto** [comment]
No, I mean `script/bench` which runs full parsing benchmarks against master and modified code.

Generally one should expect robots.txt content to be anything. Two or more asterisks is definitely possible in the wild web.

**Comment by temoto** [comment]
Also, according to this benchmark output, `TrimSuffix` is as good as slice, maybe because compiler removed it because of `_ = foo`.

**Comment by temoto** [comment]
11ns sounds pretty good to me. Maybe it should be just `TrimRightFunc` as in your first patch.

I'd think unexported module level function produce best result, but compilers will always surprise.
```go
package robotstxt
func isAsterisk(r rune) bool { return r == '*' }
```

And thanks for making this library faster. Much appreciated.

**Comment by moredure** [comment]
I guess _ = not optmized

```go
func BenchmarkTrimSuffix(b *testing.B) {
	const line = "Allow: *"
	var x string
	b.Run("HasSuffix+TrimRight", func(b *testing.B) {
		b.ReportAllocs()
		for i := 0; i < b.N; i++ {
			if strings.HasSuffix(line, "*") {
				x = strings.TrimRight(line, "*")
			}
		}
	})
	b.Run("HasSuffix+TrimRightFunc", func(b *testing.B) {
		b.ReportAllocs()
		for i := 0; i < b.N; i++ {
			if strings.HasSuffix(line, "*") {
				x = strings.TrimRightFunc(line, func (r rune) bool {
					return r == '*'
				})
			}
		}
	})
	b.Run("TrimRightFunc", func(b *testing.B) {
		b.ReportAllocs()
		for i := 0; i < b.N; i++ {
			x = strings.TrimRightFunc(line, func (r rune) bool {
				return r == '*'
			})
		}
	})
	b.Run("TrimSuffix", func(b *testing.B) {
		b.ReportAllocs()
		for i := 0; i < b.N; i++ {
			x = strings.TrimSuffix(line, "*")
		}
	})

	b.Run("TrimSuffixManually", func(b *testing.B) {
		b.ReportAllocs()
		for i := 0; i < b.N; i++ {
			if strings.HasSuffix(line, "*") {
				x = line[:len(line)-1]
			}
		}
	})
	runtime.KeepAlive(x)
}
```

```
cpu: Intel(R) Core(TM) i5-7600 CPU @ 3.50GHz
BenchmarkTrimSuffix
BenchmarkTrimSuffix/HasSuffix+TrimRight
BenchmarkTrimSuffix/HasSuffix+TrimRight-4         	23383860	        46.20 ns/op	      24 B/op	       1 allocs/op
BenchmarkTrimSuffix/HasSuffix+TrimRightFunc
BenchmarkTrimSuffix/HasSuffix+TrimRightFunc-4     	94261996	        12.20 ns/op	       0 B/op	       0 allocs/op
BenchmarkTrimSuffix/TrimRightFunc
BenchmarkTrimSuffix/TrimRightFunc-4               	95601463	        12.03 ns/op	       0 B/op	       0 allocs/op
BenchmarkTrimSuffix/TrimSuffix
BenchmarkTrimSuffix/TrimSuffix-4                  	1000000000	         0.5703 ns/op	       0 B/op	       0 allocs/op
BenchmarkTrimSuffix/TrimSuffixManually
BenchmarkTrimSuffix/TrimSuffixManually-4          	1000000000	         0.6341 ns/op	       0 B/op	       0 allocs/op
```

**Comment by moredure** [comment]
but maybe conditional assign not optimised

**Comment by moredure** [comment]
I've got another solution, maybe

```go
for z := len(x); z > 0 && x[z-1:] == "*"; z = len(x) {
   x = x[:z-1]
}
```
I guess it's most optimal and handles case with multiple "*" at the end

**Comment by moredure** [comment]
The fastest one is
```go
for z := len(x); z > 0; {
  z--
  if x[z] != '*' {
	  break
  }
  x = x[:z]
}
```

**Comment by temoto** [comment]
I think it is too complicated to understand.

**Comment by moredure** [comment]
The funny part, when I separated this code into separete code block
it becomes faster
```go
func TrimAsterisk(x string) string {
	for z := len(x); z > 0; {
		z--
		if x[z] != '*' {
			return x
		}
		x = x[:z]
	}
	return x
}
```

**Comment by moredure** [comment]
But this way it will not handle possible utf-8 runes (two bytes symbols, which may contain dot ascii code)

**Comment by temoto** [comment]
So in current project benchmark suite there is not enough `disallow: *` lines, I've run `TrimRightFunc` and custom `TrimRightAsterisk` against master.

## master vs TrimRightFunc
```
benchmark                         old ns/op     new ns/op     delta
BenchmarkParseFromString001-4     2855          2710          -5.08%
BenchmarkParseFromStatus401-4     24.5          23.7          -3.27%
BenchmarkParseFromString002-4     304489        307000        +0.82%

benchmark                         old allocs     new allocs     delta
BenchmarkParseFromString001-4     14             13             -7.14%
BenchmarkParseFromStatus401-4     0              0              +0.00%
BenchmarkParseFromString002-4     1485           1485           +0.00%
```

## master vs TrimRightAsterisk
```
benchmark                         old ns/op     new ns/op     delta
BenchmarkParseFromString001-4     2855          2742          -3.96%
BenchmarkParseFromStatus401-4     24.5          24.0          -2.04%
BenchmarkParseFromString002-4     304489        302218        -0.75%

benchmark                         old allocs     new allocs     delta
BenchmarkParseFromString001-4     14             13             -7.14%
BenchmarkParseFromStatus401-4     0              0              +0.00%
BenchmarkParseFromString002-4     1485           1485           +0.00%
```

`TrimRightFunc` is both more readable and faster. I agree needless allocation was a problem.

File: parser.go [hunk 1]
```diff
@@ -171,9 +171,7 @@ func (p *parser) parseLine() (li *lineInfo, err error) {
 			if !strings.HasPrefix(t2, "*") && !strings.HasPrefix(t2, "/") {
 				t2 = "/" + t2
 			}
-			if strings.HasSuffix(t2, "*") {
-				t2 = strings.TrimRight(t2, "*")
-			}
+			t2 = TrimRightAsterisk(t2)
```
Comment (temoto) [inline parser.go:174]:
```suggestion
			﻿t2 = strings.TrimRightFunc(t2, isAsterisk)
```

**Review by temoto — CHANGES_REQUESTED** [review]

**Comment by temoto** [comment]
Thank you very much.

**Pull request merged**